core/src/apps/cardano/sign_tx.py

```diff
@@ -117,12 +139,41 @@ def _validate_outputs(
         raise wire.ProcessError("Total transaction amount is out of range!")
 
 
+def _validate_certificates(certificates: List[CardanoTxCertificateType]) -> None:
+    for certificate in certificates:
+        if not is_staking_path(certificate.path):
+            raise INVALID_CERTIFICATE
+
+        if certificate.type == CardanoCertificateType.STAKE_DELEGATION:
+            if certificate.pool is None or len(certificate.pool) != POOL_HASH_SIZE:
+                raise INVALID_CERTIFICATE
+
+
+def _validate_withdrawals(withdrawals: List[CardanoTxWithdrawalType]) -> None:
+    for withdrawal in withdrawals:
+        if not is_staking_path(withdrawal.path):
+            raise INVALID_WITHDRAWAL
+
+        if not 0 <= withdrawal.amount < LOVELACE_MAX_SUPPLY:
+            raise INVALID_WITHDRAWAL
+
+
 def _serialize_tx(keychain: seed.Keychain, msg: CardanoSignTx) -> Tuple[bytes, bytes]:
     tx_body = _build_tx_body(keychain, msg)
     tx_hash = _hash_tx_body(tx_body)
 
-    witnesses = _build_witnesses(keychain, msg.inputs, tx_hash, msg.protocol_magic)
+    witnesses = _build_witnesses(
+        keychain,
+        msg.inputs,
+        msg.certificates,
+        msg.withdrawals,
+        tx_hash,
+        msg.protocol_magic,
+    )
 
+    # We always set transaction metadata to None, even if metadata
+    # hash is set. Metadata aren't sent to Trezor and the None
+    # should be replaced by the SW wallet if metadata exist.
```

Comment:
This sounds fishy. The user is confirming a metadata hash with no credible way to validate its value.

Why aren't the metadata sent to Trezor? Are they too complex?

If yes, do we need metadata support?

Reply:
As far as we know, Tx metadata can be up to 0.5kB and IOHK did not provide any detailed spec on it (probably it will be cbor-encoded map with binary values) + we are not sure about use cases. We can send it to Trezor but
1) it will eat up precious space for CBOR serialization
2) Trezor will not be able to display it anyway

Metadata also have no effect on spending or staking, nor do they affect key derivation, so from the hardware wallet standpoint, they do not present additional security risk. Users that are interested in verifying the metadata can actually do so (e.g. by a "manual" computation from what their software wallet is showing them). IOHK considers it necessary to have metadata support for hw wallets.

Reply:
Please send raw metadata - 512 bytes is something we can affort. We won't show them on display now, but we don't need to change the protocol once we decide to show them.

Reply:
Alright, I've done that [here](https://github.com/trezor/trezor-firmware/pull/1145/commits/1bcf5a81ebf3b0df03b25766cd7f94749233fcbf). I had to bring back `cbor.Raw`. I've tested it with 514B, some inputs and outputs and it worked no problem on a physical Trezor device. I've also tested it on the testnet.

Reply:
Transaction metadata is described in the Shelley design specification (linked from the [README](https://github.com/input-output-hk/cardano-ledger-specs)), appendix E. That contains the binary schema.

The tx body contains only the hash of the metadata (if any). The actual metadata sits outside the tx body next to the witnesses (segwit style). There is *no size limit* for the metadata, other than the overall tx size limit. Of course hardware wallets may have restrictions that prevent them producing transactions up to the maximum limit.

It is correct that the metadata has no semantic effect on the transaction on-chain. It is used to enable other applications that may use and interpret additional custom signed data attached to a transaction.